src/app/AutoResubscribeClient.h

```diff
@@ -0,0 +1,121 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+#include <app/ReadClient.h>
+#include <app/ResubscribePolicyDelegate.h>
+
+namespace chip {
+namespace app {
+class AutoResubscribeClient
+{
+public:
+    CHIP_ERROR Init(Messaging::ExchangeManager * apExchangeMgr, ReadClient::Callback * apCallback,
+                    ResubscribePolicyDelegate * apResubscribeDelegate = nullptr);
+    /**
+     *  Shut down the Client. This terminates this instance of the object and releases
+     *  all held resources.  The object must not be used after Shutdown() is called. The
+     *  ReadClient will automatically shut itself down when internal error happens.  So manual shutdown is only needed
+     *  if there is an active subscription or retrying subscription.
+     */
+    void Shutdown();
+
+    // Pass the ownership for ReadPrepareParams from application to AutoResubscribeClient, the attribute path list and event path
+    // list should be allocated in heap.
+    CHIP_ERROR SendSubscribeRequest(ReadPrepareParams && aReadPrepareParams);
+
+private:
+    class Callback final : public app::ReadClient::Callback
+    {
+    public:
+        CHIP_ERROR Init(ReadClient::Callback * apCallback, AutoResubscribeClient * apAutoReSubscribeClient)
+        {
+            VerifyOrReturnError((apCallback != nullptr) && (apAutoReSubscribeClient != nullptr), CHIP_ERROR_INVALID_ARGUMENT);
+            mpCallback              = apCallback;
+            mpAutoReSubscribeClient = apAutoReSubscribeClient;
+            return CHIP_NO_ERROR;
+        }
+
+        void Reset()
+        {
+            mpCallback              = nullptr;
+            mpAutoReSubscribeClient = nullptr;
+        }
+
+    private:
+        void OnReportBegin(const ReadClient * apReadClient) override { mpCallback->OnReportBegin(apReadClient); }
+
+        void OnReportEnd(const ReadClient * apReadClient) override { mpCallback->OnReportEnd(apReadClient); }
+
+        void OnEventData(const ReadClient * apReadClient, const EventHeader & aEventHeader, TLV::TLVReader * apData,
+                         const StatusIB * apStatus) override
+        {
+            mpCallback->OnEventData(apReadClient, aEventHeader, apData, apStatus);
+        }
+
+        virtual void OnAttributeData(const app::ReadClient * apReadClient, const app::ConcreteDataAttributePath & aPath,
+                                     TLV::TLVReader * apData, const app::StatusIB & aStatus) override
+        {
+            mpCallback->OnAttributeData(apReadClient, aPath, apData, aStatus);
+        }
+
+        void OnError(const app::ReadClient * apReadClient, CHIP_ERROR aError) override
+        {
+            mpCallback->OnError(apReadClient, aError);
```

Comment:
This is not going to work - the moment `OnError` is called by the underlying ReadClient object, it's on its way to being shutdown and should not be used further.

Reply:
Per offline chat with @bzbarsky-apple @mrjerryjohns , decides to put resubscribe logic inside readClient. update the PR, thanks